File: myfragments/src/main/java/pl/mareklangiewicz/myfragments/MyPieTestsFragment.java
package pl.mareklangiewicz.myfragments;


import android.animation.ObjectAnimator;
import android.animation.PropertyValuesHolder;
import android.graphics.Color;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.ColorInt;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AccelerateInterpolator;

import pl.mareklangiewicz.myutils.MyMathUtils;
import pl.mareklangiewicz.myviews.IMyNavigation;
import pl.mareklangiewicz.myviews.MyPie;

//TODO LATER: create UI to change animation speed (in local menu header)

public final class MyPieTestsFragment extends MyFragment implements View.OnClickListener {

    private @NonNull String mRandomize = "to";

    private @Nullable ObjectAnimator mHeaderAnimator;

    public MyPieTestsFragment() { }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.my_pie_tests_fragment, container, false);

        inflateHeader(R.layout.my_pie_tests_header);
        inflateMenu(R.menu.my_pie_tests_menu);

        root.findViewById(R.id.pie1).setOnClickListener(this);
        root.findViewById(R.id.pie2).setOnClickListener(this);
        root.findViewById(R.id.pie3).setOnClickListener(this);
        root.findViewById(R.id.pie4).setOnClickListener(this);

        //noinspection ConstantConditions
        MyPie hpie = (MyPie) getHeader().findViewById(R.id.header_pie);

        float min = hpie.getMinimum();
        float max = hpie.getMaximum();

        PropertyValuesHolder pvh1 = PropertyValuesHolder.ofFloat("to", min, max);
        PropertyValuesHolder pvh2 = PropertyValuesHolder.ofFloat("from", min, max - (max - min) / 2);

        mHeaderAnimator = ObjectAnimator.ofPropertyValuesHolder(hpie, pvh1, pvh2);

        mHeaderAnimator.setInterpolator(new AccelerateInterpolator());

        if(savedInstanceState == null)
            selectItem(R.id.mpt_randomize_to);

        return root;
    }


    @Override
    public void onDestroyView() {
        mHeaderAnimator = null;
        super.onDestroyView();
    }

    @Override
    public void onResume() {
        super.onResume();
        MenuItem item = getFirstCheckedItem(); //see MyNavigationView.getFirstCheckedItem warning..

        if(item == null) {
            log.e("No item selected");
            mRandomize = "to";
        }
        else
            mRandomize = item.getTitle().toString();
    }

    @Override
    public boolean onItemSelected(IMyNavigation nav, MenuItem item) {
        if(nav != getLocalNavigation()) {
            log.v("This item is not from our local menu.");
            return false;
        }
        mRandomize = item.getTitle().toString();
        return true;
    }

    @Override
    public void onClick(View v) {
        if(v instanceof MyPie) {
            MyPie pie = (MyPie) v;
            if(mRandomize.equals("pieColor") || mRandomize.equals("ovalColor")) {
                @ColorInt int value = MyMathUtils.getRandomColor(Color.rgb(0, 0, 0), Color.rgb(255, 255, 255));
                if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                    ObjectAnimator.ofArgb(pie, mRandomize, value).start();
                    log.i("[SNACK]MyPie:%s: random color is ... %X %X %X", mRandomize, Color.red(value), Color.green(value), Color.blue(value));
                }
                else
                    log.w("[SNACK]Color animation is not supported on platforms before Lollipop");
            }
            else {
                float min = 0;
                float max = 100;
                switch(mRandomize) {
                    case "from":
                        min = pie.getMinimum();
                        max = pie.getTo();
                        break;
                    case "to":
                        min = pie.getFrom();
                        max = pie.getMaximum();
                        break;
                    case "minimum":
                        max = pie.getFrom();
                        break;
                    case "maximum":
                        min = pie.getTo();
                        break;
                }

                float value = MyMathUtils.getRandomFloat(min, max);
                ObjectAnimator.ofFloat(pie, mRandomize, value).start();
                log.i("[SNACK]MyPie:%s: random %.2f..%.2f is ... %.2f", mRandomize, min, max, value);
            }
        }
    }


    @Override
    public void onDrawerSlide(View view, float slideOffset) {
        if(view != getLocalNavigation())
            return;

        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP_MR1) {
            if(mHeaderAnimator != null)
                mHeaderAnimator.setCurrentFraction(slideOffset);
        }
    }

    @Override public void onDrawerOpened(View view) { }

    @Override public void onDrawerClosed(View view) { }

    @Override public void onDrawerStateChanged(int i) { }
}

File: myfragments/src/main/java/pl/mareklangiewicz/myfragments/MyPieTestsFragment.kt
package pl.mareklangiewicz.myfragments


import android.animation.ObjectAnimator
import android.animation.ObjectAnimator.ofArgb
import android.animation.PropertyValuesHolder
import android.graphics.Color
import android.graphics.Color.*
import android.os.Bundle
import androidx.annotation.ColorInt
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AccelerateInterpolator
import kotlinx.android.synthetic.main.mf_my_pie_tests_fragment.*
import kotlinx.android.synthetic.main.mf_my_pie_tests_header.view.*
import pl.mareklangiewicz.myutils.*
import pl.mareklangiewicz.myviews.MyPie
import java.lang.String.format

//TODO SOMEDAY: create UI to change animation speed (in local menu header)

open class MyPieTestsFragment : MyFragment(), View.OnClickListener {

    private val tocancel = Lst<(Cancel) -> Unit>()

    private var randomize = "to"

    private lateinit var hanimator: ObjectAnimator

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState) //just for logging
        return inflater.inflate(R.layout.mf_my_pie_tests_fragment, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        manager?.lnav?.headerId = R.layout.mf_my_pie_tests_header
        manager?.lnav?.menuId = R.menu.mf_my_pie_tests

        mf_mptf_pie1.setOnClickListener(this)
        mf_mptf_pie2.setOnClickListener(this)
        mf_mptf_pie3.setOnClickListener(this)
        mf_mptf_pie4.setOnClickListener(this)

        manager?.lnav?.headerObj?.let {
            val min = it.mf_mpth_pie.minimum
            val max = it.mf_mpth_pie.maximum
            val pvh1 = PropertyValuesHolder.ofFloat("to", min, max)
            val pvh2 = PropertyValuesHolder.ofFloat("from", min, max - (max - min) / 2)
            hanimator = ObjectAnimator.ofPropertyValuesHolder(it.mf_mpth_pie, pvh1, pvh2)
            hanimator.interpolator = AccelerateInterpolator()
        }

        val ctl = manager!!.lnav!!.items {
            randomize = manager!!.lnav!!.menuObj!!.findItem(it)?.toString() ?: randomize
        }
        tocancel.add(ctl)

        if (savedInstanceState == null)
            manager?.lnav?.setCheckedItem(R.id.mpt_randomize_to, true)
    }

    override fun onResume() {
        super.onResume()
        randomize = manager?.lnav?.firstCheckedItem?.run { title.toString() } ?: "to".apply { log.e("No item selected") }
    }

    override fun onDestroyView() {
        tocancel.forEach { it(Cancel) }
        tocancel.clr()
        manager?.lnav?.headerId = -1
        manager?.lnav?.menuId = -1
        super.onDestroyView()
    }

    override fun onClick(v: View) {

        if ( v !is MyPie) return

        if (randomize == "pieColor" || randomize == "ovalColor") {
            @ColorInt val value = RANDOM.nextColor(Color.rgb(0, 0, 0), Color.rgb(255, 255, 255))
            ofArgb(v, randomize, value).start()
            log.i(format("[SNACK]MyPie:%s: random color is ... %X %X %X", randomize, red(value), green(value), blue(value)))
        } else {
            var min = 0f
            var max = 100f
            when (randomize) {
                "from" -> {
                    min = v.minimum
                    max = v.to
                }
                "to" -> {
                    min = v.from
                    max = v.maximum
                }
                "minimum" -> max = v.from
                "maximum" -> min = v.to
            }

            val value = RANDOM.nextFloat(min, max)
            ObjectAnimator.ofFloat(v, randomize, value).start()
            log.i(String.format("[SNACK]MyPie:%s: random %.2f..%.2f is ... %.2f", randomize, min, max, value))
        }
    }


    override fun onDrawerSlide(drawerView: View, slideOffset: Float) {
        if (drawerView !== manager?.lnav) return
        hanimator.setCurrentFraction(slideOffset)
    }
    override fun onDrawerOpened(drawerView: View) { }
    override fun onDrawerClosed(drawerView: View) { }
    override fun onDrawerStateChanged(newState: Int) { }
}
